### ISSUE TITLE
BUG: numpy.ndarray.__deepcopy__ returned a result with an exception set

### ISSUE THREAD

**Issue by mdickinson** [numpy/numpy#21883]
### Describe the issue:

It looks as though there's a code path in which `ndarray.__deepcopy__` sets an exception but fails to return `NULL`, confusing Python as a result. This was reported upstream on the CPython bug tracker at https://github.com/python/cpython/issues/94390, but looks as though it's likely a problem with NumPy.

#8706 looks relevant here. Specifically, I suspect that `itemp` is `NULL` in [this](https://github.com/numpy/numpy/blob/4d057e633408c04c15292ccc9cd11d96e33a90f5/numpy/core/src/multiarray/methods.c#L1636) call, resulting in the `PyObject_CallFunctionObjArgs` argument list being cut short.

I wasn't able to reproduce with an empty array of dtype object created directly with `np.empty`; it's not clear to me when such an array gets filled with NULLs and when it gets filled with pointers to `Py_None`.

### Reproduce the code example:

```python
import numpy as np
np.empty_like([slice(3)]).__deepcopy__({})
```

### Error message:

```shell
TypeError: deepcopy() missing 1 required positional argument: 'x'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <method '__deepcopy__' of 'numpy.ndarray' objects> returned a result with an exception set
```

### NumPy/Python version information:

1.22.4 3.10.5 (main, Jun  7 2022, 08:51:33) [Clang 13.0.0 (clang-1300.0.29.30)]

**Comment by seberg** [comment]
Can't reproduce on `main`, but that is just because we started filling empty-like arrays with `None` (for the time being), rather than `NULL`.  Branches where `NULL` happens are still plausible (if very esoteric), so this needs fixing.

The analysis is thus perfectly correct, we should use `None` here when the value is `NULL` to be robust.  Seems like a simple enough fix.

**Comment by seberg** [comment]
Then this is fixed, gh-8706 should probably just be fixed in one go.

**Comment by jcusick13** [comment]
I'm at the Scipy sprints and will try taking a stab at this today

**Comment by cmarmo** [comment]
Hello, while searching for some issues to tackle I found this one.... if I understand correctly it could be closed as solved by #21996?

**Comment by seberg** [comment]
Yes, thought it would get auto-closed there, thanks.
